io-engine/src/core/block_device.rs

```diff
@@ -170,30 +158,184 @@ pub trait BlockDeviceHandle {
+            .await
+    }
+
+    /// Writes the given number of blocks from the list of buffers to the
+    /// device, starting at the given offset.
+    ///
+    /// The caller must ensure that the number of blocks to write does not go
+    /// beyond the size of `iovs` array.
+    ///
+    /// The given completion callback is called when the operation finishes.
+    /// This method may return error immediately in the case operation dispatch
+    /// fails.
     fn writev_blocks(
         &self,
-        iov: *mut IoVec,
-        iovcnt: i32,
+        iovs: &[IoVec],
         offset_blocks: u64,
         num_blocks: u64,
         cb: IoCompletionCallback,
         cb_arg: IoCompletionCallbackArg,
     ) -> Result<(), CoreError>;
 
+    /// Writes the given number of blocks from the list of buffers to the
+    /// device, starting at the given offset.
+    ///
+    /// The caller must ensure that the number of blocks to write does not go
+    /// beyond the size of `iovs` array.
+    ///
+    /// Operation is performed asynchronously; I/O completion status is wrapped
+    /// into `CoreError::WriteFailed` in the case of failure.
+    async fn writev_blocks_async(
+        &self,
+        iovs: &[IoVec],
+        offset_blocks: u64,
+        num_blocks: u64,
+    ) -> Result<(), CoreError> {
+        let (s, r) = oneshot::channel::<IoCompletionStatus>();
+
+        self.writev_blocks(
+            iovs,
+            offset_blocks,
+            num_blocks,
+            block_device_io_completion,
+            cb_arg(s),
+        )?;
+
+        match r.await.expect("Failed awaiting at writev_blocks()") {
+            IoCompletionStatus::Success => Ok(()),
+            status => Err(CoreError::WriteFailed {
+                status,
+                offset: offset_blocks,
+                len: num_blocks,
+            }),
+        }
+    }
+
+    /// Writes the given number of blocks from the buffer to the device,
+    /// starting at the given offset.
+    ///
+    /// The caller must ensure that the `buf` buffer is large enough to write
+    /// `num_blocks`.
+    ///
+    /// Operation is performed asynchronously; I/O completion status is wrapped
+    /// into `CoreError::WriteFailed` in the case of failure.
+    async fn write_buf_blocks_async(
+        &self,
+        buf: &DmaBuf,
+        offset_blocks: u64,
+        num_blocks: u64,
+    ) -> Result<(), CoreError> {
+        let iovs = [unsafe { buf.as_iovec() }];
+        self.writev_blocks_async(&iovs, offset_blocks, num_blocks)
+            .await
+    }
+
+    /// TODO
+    async fn verify_buf_blocks_async(
+        &self,
+        src_buf: &DmaBuf,
+        offset_blocks: u64,
+        num_blocks: u64,
+    ) -> Result<bool, CoreError> {
+        let mut dst_buf = self.dma_malloc(src_buf.len()).map_err(|_| {
+            CoreError::DmaAllocationFailed {
+                size: src_buf.len(),
+            }
+        })?;
+
+        self.read_buf_blocks_async(
```

Comment:
Can't we use compare command? This way we don't need to receive any data and the target will compare it for us.

Reply:
Yes we can. I try to use it.

Reply:
Done. I had to made a fix to SPDK to make compare APIs work properly: https://github.com/openebs/spdk/pull/35.

Reply:
Yes I saw, approved that one already.